### ISSUE TITLE
Discussion: Options for Authenticator Decoupling

### ISSUE THREAD

**Issue by jonahx** [cyberark/conjur#2355]
# Options for Authenticator De-coupling

## Overview

There are two main approaches we can take to de-coupling the authenticators:

1. **Full decoupling** -- Make each authenticator an independent web service. 
  - Necessarily, access to Conjur is only via the API.  
  - Writes directly to Audit log socket for audited events.
  - Non-audit logs are separate from Conjur's.
  - Requires configuration (possibly automated) in nginx (or whichever webserver you use).  Authenticator services define their own routes, which are mounted under a standard path.  For example: `authn/k8s` or `authn/oidc`.
  - Existing authenticators can be ported to simple sinatra of rack apps.  Additional code will be minimal: Just the route definitions, which will then point to the existing code unchanged.
  - Conjur will provide tokens via authn-local.
2. **Partial decoupling** -- Make each authenticator a gem, but still execute code within Conjur's process.
  - Authenticator provides routes, but Conjur code writes those routes into existing `config/routes.rb` file.
  - Authenticator constructors accept objects giving access to rails logger, audit logger, and interfaces to relevant database queries.  These latter need to be designed.

### Pros

- Long-term, imo this is probably the right solution.
- Restricting Conjur access through API is cleaner and more maintainable.
- Enables fully pluggable 3rd party authenticators, as requested by the field and sales, should we decide to allow that (eg, via `UNSAFE` flags or similar).
- Can be written in any language.

### Cons

- Requires configuration (even if automated) of the webserver itself, since each authenticator will run as its own service.  We can't mount them as middleware within Conjur, because they need to make requests to conjur themselves.
- While audit logs can write to the same socket that Conjur does, the individual services will need their own logs, or we'll need an aggregation solution.  This means more files to check when debugging.
- Security is (arguably) harder.  We already have the active `authn-local` which mints tokens, so even right now anyone _could_ write their own fully pluggable authenticator and configure the webserver to serve it.  All we'd be doing is using that existing functionality officially ourselves.  The security model (which, again, we already implicitly have) is essentially "anyone who is able to configure a new service on the web server has the authority to mint tokens."

## Discussion

These are notes based on a conversation with Matthew Brace.  Like me, he strongly prefers this approach _if_ we can make it secure.  One requirement for security is to protect the authn-local socket with authentication of its own, so that if the box itself is owned it won't immediately mean that all Conjur secrets are owned. One promising idea, which could also simplify configuration of the services, is to: 

1. Use rack's middleware proxy in Conjur to route requests to the separate-process authenticator services.
2. Inject a short-lived, one time use authentication token that can then be used by the authenticator to access the protected authn-local socket which provides tokens.
3. Use SSL between Conjur and the authenticator services.

### Pros

- Still have just a single Conjur service.
- No messing with web server configurations.
- Security is simpler since new authenticators can only be enabled by an explicit PR into the Conjur repo.  No possibility of a non-vetted authenticator ever being used.

### Cons

- Not truly decoupled.  Since we are providing non-API database access via objects, a new authenticator that requires a query we don't already provide will require a new database access object to be written, and thus a separate PR.
- Even when that's not the case, updates to authenticators will always require both the PR on the authenticator gem and a separate PR on Conjur to bring in the update.  This makes development less convenient.
- Authenticators can only be written in ruby.

**Comment by dataplex** [comment]
Can this be moved to a sharepoint document so it has a wider audience?

**Comment by dataplex** [comment]
What is the timeline for this being worked on? Is this on the schedule already, and are there constraints here that need to drive our suggestions? I'm concerned that neither of these options expand on the operational impact, meaning the actual day to day use of these options versus the current system is not well defined. Given that, I'll do my best to provide feedback as is and hopefully we can clarify as we go.

### Operational Issues - Both Solutions

1. Service Restarts at Scale
    1. Currently we have to restart the 'conjur' services on each Conjur instance that is affected by enabling an authenticator.
    2. If neither of these solutions address that scalability issue, we should see if we can come up with a solution to that problem. (see below)
2. Authenticator Configuration
    1. Currently configuration of authenticators is two fold: 1) create the authenticator policy, 2) enable the authenticator in <env, config.yml>.
    2. Related to the above, but also complicated by both options presented.
3. Logging
    1. Open ended logging of authenticators separate from the conjur service through which the requests are sent would require complex log file parsing to create a full view of a request from authentication, through request. That will make troubleshooting extremely difficult. The field needs to see end to end request start to request end, including any authentication that occurs. These logs need to be sequential and time based. It appears that at least one of these options would require separate process IDs, making correlation at scale practically impossible.
    2. Having seperate log files for authenticators means an inability to define a finite set of log files that need to be monitored. It also means configuring log file rotation changes over time, and if that is not included by default, could lead to a loss of service as a result of storage resource exhaustion.
    3. Standardization - Authenticators should, in general, have common logging patterns to make it easier to troubleshoot the system as a whole. Neither of these solutions appear to address log message standardization.


### Full Decoupling

1. General Feedback
    1. This may be more maintainable from a development experience standpoint, but operationally this option - as presented - appears much more complex. This is pushing the burden of complexity onto the customer rather that at the expertise point (R&D).
    2. Requires many more touch points to many internal service configurations - nginx, log rotator, database access, certificates, syslog configs, etc.
    3. Security, availability, etc, implications of conflicting configurations, config changes that overwrite other important configurations, etc.
    4. With this option we lose the ability to ensure a quality user experience, and given the complexity of the Conjur environment, it is almost certain that this would lead to a fragmented experience.
2. "Restricting Conjur access through API is cleaner and more maintainable."
    1. Please expand on this statement. Does this mean that this option would change the available APIs based on what is enabled?
    2. Assuming the above is true, this would make creating comprehensive SDKs (client libraries) nearly impossible. Extending authenticators in this manner would also mean additional work creating and documenting extension points in all supported client libraries.
    3. Allows for situations where an authenticator is enabled on the host but not configured in policy. How does this influence service discovery?
3. "Writes directly to Audit log socket for audited events."
    1. This would mean that 3rd party authenticators can't leverage the assigned IANA Private Enterprise Number (PEN). 
    2. Technically they aren't part of the Conjur system, so we would have an inconsistent audit log experience.
    3. This doesn't enforce log message format, meaning that parsing audit logs could result in failure.
    4. Also creates an inconsistent experience regarding log level, source, message format, etc.
    5. Would this write the messages to both the Audit UI database and the log files?
4. "Non-audit logs are separate from Conjur's."
    1. Log rotation concerns, disk space management, no ability to give finite log file monitoring recommendations.
    2. See general comments about disparate troubleshooting experience.
5. "Requires configuration (possibly automated) in nginx (or whichever webserver you use). ...."
    1. Are we suggesting running multiple web servers in the container?!
    2. How does this work with backup and restore? Would we require the backup to store the software, and configuration, and be able to restore that?
    3. Or would this be implemented as separate containers? I like this idea!!!
6. "Existing authenticators can be ported to simple sinatra of rack apps."
    1. Exponentially increases the complexity of service dependency management, service restarts, backup/restore, etc.
    2. If there is unstable code in an authenticator, it could crash independently of the Conjur service, making service integrity and availability difficult to monitor.

### Partial Decoupling
1. "Authenticator provides routes, but Conjur code writes those routes into existing config/routes.rb file."
   1. How would route discovery take place?
   2. Supply chain attack concerns - how do we offer package integrity validation?
   3. Can we easily detect and troubleshoot conflicting routes, etc. Wouldn't load order allow for situations in which one authenticator could accidentally hijack requests  from another authenticator? This appears like it would be a runtime error and not a deploy-time / configuration error.
2. "Authenticator constructors accept objects giving access to rails logger, audit logger, and interfaces to relevant database queries. These latter need to be designed."
    1. Is this truly decoupled?
    2. This would require coordinating changes to any of these dependency interfaces across all popular authenticators (including custom ones written by customers, a very likely thing to spring up out of this effort).
    3. This seems very difficult, and we would have to understand all the potential data extensions required by authenticators, a likely impossible task.

### Suggestion - Fully Decoupled Solution with Conjur Platform Core Services

While the partially decoupled solution would offer a more consistent experience, it also creates a version of dependency hell I think we would wish to avoid. The fully decoupled solution seems more promising while also suffering from some severe user experience and operational concerns.

I would like to propose the following, addressing both operational and development concerns:

1. Identify required core services and create separate [RPC] services exposing these common components:
    1. Audit Logging
    2. Database access - provide a standard means of accessing the database that provides audit logging of the database activity in addition to authenticator specific audit logging
    3. Message versioning - Allow authenticator authors time to update their libraries when the interfaces change. This can be accomplished either by exposing multiple endpoints (separate rack services) or by introducing message versioning. I would vote for multiple endpoints as this would allow for long term separation of these services from the big Conjur container later on, making development of the core even more simplified.
    4. Bonus: Provide an authenticator client library for consistent access to the above

2. Authenticator registration
    1. registration via configuration
    2. Kisses
    3. Explicit definition of core service requirements
    4. Ability to create database space (if requested) for the authenticator that does not expose the entire Conjur database (yes, this is a bit of a dog and pony show, but legitimate authors will adhere to it)
    5. Ability to extend authenticator installation with integrity validation
3. Authenticator discovery
    1. Configuration file based registration of authenticators allows for simplified discovery of available endpoints. Allows explicit deployment of authenticator code.
4. Requirement to do all other configuration via policy
    1. Only configuration that can be validated at start time should be done via config file.
    2. Allow for scaling and true automation based management
5. SINGLE /authenticate endpoint
     1. Restructured authenticate message body to allow for a SINGLE route that can then delegate authentication requests based on the registered authenticators that can handle that message
     2. Easy to test - register authenticators, fire a message, catch the audit.
     3. Allows for an authentication experience that is extensible, discoverable, and can be validated at startup rather than runtime. DOES NOT require service restarts to enable authenticator configurations!
6. Validation endpoint
    1. Verify whether a defined resource is configured correctly for a particular authenticator. Delegates the burden of implementation detail verfiication from the COnjur admin to the owning team/individual
    2. Easily determine whether that resource can authenticate or not
    3. Bonus idea: Allow registration of resource creation handlers so that validation can be done at resource creation time rather than consumer runtime. More detailed and actionable error messages, earlier in the process.

### Example

config.yml:
```
coreservices:
  audit:
    loglevel: warn
    requires:
      facility: 1
      messages:
        authentication_successful
          severity: 5
        authentication_failure
          severity: 5
        authentication_trace
          severity: 7
    latest: audit@1.2.4
    audit@1.2.3:
      endpoint:
        type: unix_socket
        uri: /var/run/audit_1-2-3/audit.sock
    audit@1.2.4:
      endpoint:
        type: http
        uri: localhost:54533
timeouts:
  registration: 10s
  heartbeat: 2s
  retry: 5s
  max_retries: 3
authenticators:
  - name: authn-k8s
    latest: authn-k8s@1.2.4
    authn-k8s@1.2.4:
      active: yes
      type: gem
      gem:
        - conjur-authn-k8s@1.2.4
        - validate: sha1
        - entry_class: AuthnK8sRequestHandler
      requirements: {{PATH}}/requirements.yml
      handles:
        messages:
          type: authn-k8s/*
          versions:
            - 1.2.4
            - latest
      runtime:
        database:
          isolation:
            type: schema
            schema: authnk8s
            client_cert: /path/to/service/client.pem
            client_key: /path/to/service/client.key
          core:
            schema: authn_generic
              - authenticator_config
              - roles
            schema: authn_annotations
              - annotations
    authn-k8s@1.2.0:
      active: deprecated
      type: rpc
      rpc:
        endpoint:
          type: socket
          uri: /var/run/authenticators/authn-k8s/authenticator.sock
      requirements: /etc/conjur/config/authn-k8s_1-2-0_requirements.yml
      handles:
        messages:
          type: authn-k8s/*
          version: 1.2.0
      runtime:
        database:
          isolation:
            type: schema
            schema: authnk8s
            client_cert: /path/to/service/client.pem
            client_key: /path/to/service/client.key
          core:
            schema: authn_generic
              - authenticator_config
              - roles
            schema: authn_annotations
              - annotations
```
authn-k8s@1.2.4 requirements.yml
```services:
  audit: audit@latest
  database: database@1.1.0
registration:
  database:
    isolated: yes
    core_views:
      - authenticator_config
      - roles
      - annotations
```
authn-k8s@1.2.0 requirements.yml

```services:
  audit: audit@latest
  database: database@1.1.0
registration:
  database:
    isolated: yes
    core_views:
      - authenticator_config
      - roles
 ```

**Comment by infamousjoeg** [comment]
I'm not going to reply from Ben's perspective, but from the perspective of the customers I encounter in the Southeast region of the US.

They barely understand it as it stands today. All of Ben's "Fully Decoupling - General Feedback" section is completely spot on from my user's perspective.

If this was started from feedback from DevOps SMEs, I'd like it to be discussed with them prior to any further action on this to confirm the experience you're accepted from an R&D perspective is the same as their customers.

**Comment by jonahx** [comment]
These are some comment @rafis3 sent me offline.  I am responding to them here for visibility.

> I am missing your opinion, which is very important to me. What is your thought and recommendation and why?

I think there are definitely tradeoffs, and agree with some of your concerns, but as I said fully decoupled strikes me as the superior Long-term solution.  I will elaborate below...

> Originally, we discussed that we need to understand if we want to go the route of making the authenticators pluggable or if the effort, code complexity, the decoupling process is too hard, to make it worth it. 

Imo it needs to be done, if for no other reason than that it is a requirement to reduce our CI times to something reasonable.  Our current CI is ~1hr.  It's an absolute productivity killer.

> If we created partial plugins based on an engine, rather than a gem, we could have had the routes in the plugin as well, right?

Yes, but I would strongly prefer to avoid engines.  They bring in all the complexity of a complete rails app, and couple our authenticator code to rails, for no reason.  The authenticator logic has nothing to do with rails, and taking on that dependency means that: It is _yet another_ tool for developers learn (engines have subtle differences from rails); yet another thing that will need updates and might have CVE; the code will be more bloated than it needs to be.  Also, as compared with vanilla middleware, they don't change anything from the POV the problems I raised.


> The full decoupling might have performance implications, because there will be more networking between services. Kind of a step back to how Conjur v4 worked. Might not be significant, but still a risk I think The footprint will probably be larger with the full decoupling, if running more processes, more web services, because each brings its own stack which is an overhead

This is genuine concern I agree with.  Unless we're very careful, it is also likely to add to complexity.  Imo these are serious concerns that we need to discuss and find good solutions for.  I think it is probably possible to do so.

> Long-term, imo this is probably the right solution. this is a subjective “pro” as opposed to the others which are technical. So I think it’s worth adding why it’s the right solution, someone else could think differently if no explanation is provided.

My reasons are:

- It enables 3rd party development, which the field and sales engineers have been requresting for a long time.  I understand the security concerns around this but believe they can be addressed with appropriate UNSAFE flags and warnings.
- Allows authenticators to be written in any language.
- Allows authenticators to scale independently of Conjur.
- It's conceptually simpler (though, as you noted, operationally more complex).

**Comment by dataplex** [comment]
> > The full decoupling might have performance implications, because there will be more networking between services. Kind of a step back to how Conjur v4 worked. Might not be significant, but still a risk I think The footprint will probably be larger with the full decoupling, if running more processes, more web services, because each brings its own stack which is an overhead

If we had performance metrics, and baseline SLAs for API queries, I would consider this with more weight. Right now, we have neither of those things, and any discussion about performance impacts is incomplete without them. If we could establish what those are currently, we could very easily evaluate both options (and mine if inclined) with scaffolding work to determine if performance is actually negatively impacted.

That being said, I would argue that right now this product needs as much dev time on code as possible, so I would accept performance hits for short-medium length if it gained enough dev time back to actually address some core issues. In the end, we dictate to the customer the resource requirements and infrastructure for this product, meaning we can beef those up for a while if needed to address this issue. Decoupling authenticators, and considering my suggestions for restructuring core services, would allow us to develop authenticators in compiled languages much faster than Ruby, so long term performance would benefit.

**Comment by rafis3** [comment]
> Imo it needs to be done, if for no other reason than that it is a requirement to reduce our CI times to something reasonable. Our current CI is ~1hr. It's an absolute productivity killer.

There is more than one way to improve the CI. We can trigger tests that run as part of a PR and commit, based on the area that you touched. The authenticators tests don't have to run always, but can give an asynchronous feedback later. Since they also run in parallel, adding more authenticators doesn't necessarily mean more time with the important assumption that our tests are stable. We must have stable tests, so that more tests, doesn't increase the chance of random failures.

**Comment by jonahx** [comment]
@rafis3 Fwiw, after a very long discussion with Ben, he's convinced me that the impact of option 1 (at least the version described above) is not feasible for field operations, and so it's off the table for now.  We only want to find a solutions that works for everyone.

**Comment by szh** [comment]
I'm going to close this as stale. Please re-open if this is still relevant.
